Bootstrapped databases with best practice role setup

FxKu opened the pull request:
At the moment, users are left alone with setting up read and write privileges in PostgreSQL. A best practice setup could look like this:

- three NOLOGIN roles: a DB owner, a DB reader and DB writer
- [default access privileges](https://www.postgresql.org/docs/12/sql-alterdefaultprivileges.html) for these three roles
- three (optional) LOGIN roles which inherit the privileges from the NOLOGIN roles and can be granted to application users

This PR adds a new section `preparedDatabases` to the manifest as the old `databases` field can not be changed without breaking changes. The whole role setup is hidden from the user. Application roles would still be specified under the `users` field. A few more features are currently included, but can be removed as it adds quite some complexity:

- If the preparedDatabases section is empty, create a database which is named like the Postgres cluster manifest
- Allow users to define database schemas for each bootstrapped database (default schema would be called `data`)
- Optional NOLOGIN and LOGIN roles per schema
- For each preparedDatabase allow a string map to specify extensions and which schema they should run (addresses #490)

As a core principle, I tried to only use go maps of complex objects to allow for non-breaking future additions. But in general, we do not aim for adding more complexity. The operator should not become a database migration tool where you specify any DDL stuff. E.g. schemas and extension might already be too much.

sdudoladov commented on manifests/minimal-postgres-manifest.yaml:
```diff
@@ -15,5 +15,7 @@ spec:
   databases:
     foo: zalando  # dbname: owner
+  preparedDatabases:
+    bar: {}
```
does this have the effect of creating the bar DB owned by `db_owner` ?

FxKu replied:
The roles are always called like the DB. So it would be `bar_owner` etc.
Plus, there will be a schema called `data`

sdudoladov commented on pkg/cluster/sync.go:
```diff
@@ -536,6 +555,109 @@ func (c *Cluster) syncDatabases() error {
+
+	if createPreparedSchemas, equal := util.SubstractStringSlices(schemas, currentSchemas); !equal {
+		for _, schemaName := range createPreparedSchemas {
+			owner := "_owner"
```
these `_owner` / `_writer` etc.  suffixes look like constants everywhere in the code

sdudoladov commented on pkg/cluster/database.go:
```diff
@@ -224,6 +250,104 @@ func (c *Cluster) databaseNameOwnerValid(datname, owner string) bool {
+	return nil
+}
+
+func (c *Cluster) execAlterGlobalDefaultPrivileges(owner, rolePrefix string) error {
```
what if a role with the same name but different set of default privileges already exists ?

FxKu replied:
I guess Postgres will overwrite existing privileges as it's an ALTER command

sdudoladov commented on pkg/cluster/database.go:
```diff
@@ -224,6 +250,104 @@ func (c *Cluster) databaseNameOwnerValid(datname, owner string) bool {
+		schemaName, rolePrefix+"_reader", // tables
+		schemaName, rolePrefix+"_reader", // sequences
+		schemaName, rolePrefix+"_writer", // tables
+		schemaName, rolePrefix+"_writer", // sequences
```
comments mention `// sequences` twice

FxKu replied:
have a look at the SQL command. First SELECT on tables and sequences is granted to the `_reader` role, than the rest to the `_writer` role

sdudoladov commented on pkg/util/users/users.go:
```diff
@@ -74,25 +74,43 @@ func (strategy DefaultUserSyncStrategy) ProduceSyncRequests(dbUsers spec.PgUserM
 
 // ExecuteSyncRequests makes actual database changes from the requests passed in its arguments.
 func (strategy DefaultUserSyncStrategy) ExecuteSyncRequests(reqs []spec.PgSyncUserRequest, db *sql.DB) error {
+	var rr []spec.PgSyncUserRequest
```
does  `rr` stand for  `request retries` ?

sdudoladov commented on pkg/util/users/users.go:
```diff
@@ -74,25 +74,43 @@ func (strategy DefaultUserSyncStrategy) ProduceSyncRequests(dbUsers spec.PgUserM
+	// retry adding roles as long as the number of failed attempts is shrinking
+	if len(rr) > 0 {
+		if len(rr) < len(reqs) {
+			if err := strategy.ExecuteSyncRequests(rr, db); err != nil {
```
that recursive call to `ExecuteSyncRequests` is definitely worth a unit test

sdudoladov commented:
> Application roles would still be specified under the users field

With this change, what would be the preferred way to create application roles  ? With the old `users` section or with the new login roles created in this PR ?

FxKu commented on pkg/apis/acid.zalan.do/v1/postgresql_type.go:
```diff
@@ -77,6 +78,19 @@ type PostgresqlList struct {
+
+// PreparedSchema describes elements to be bootstrapped in the schema
+type PreparedSchema struct {
+	DefaultRoles *bool `json:"defaultRoles,omitempty" defaults:"true"`
```
Roles = NOLOGIN roles = safe
Users = LOGIN roles = should be enabled explicitly

FxKu commented on manifests/complete-postgres-manifest.yaml:
```diff
@@ -21,6 +21,17 @@ spec:
+        pg_partman: public
+        pgcrypto: public
+      schemas:
+        data: {}
```
yes, it's not necessary here. But we also use default in configmap etc. so it's just to show the users how to write schemas

FxKu commented on pkg/cluster/sync.go:
```diff
@@ -541,26 +547,138 @@ func (c *Cluster) syncDatabases() error {
-		currentOwner, exists := currentDatabases[datname]
+	// if no prepared databases are specified create a database named like the cluster
+	preparedDatabases := c.Spec.PreparedDatabases
+	if preparedDatabases == nil || len(preparedDatabases) == 0 { // TODO: add option to disable creating such a default DB
```
should be fixed now

FxKu commented on pkg/cluster/sync.go:
```diff
@@ -541,26 +547,138 @@ func (c *Cluster) syncDatabases() error {
+		}()
+
+		// first, set default privileges for prepared database
+		c.execAlterGlobalDefaultPrivileges(preparedDbName+constants.OwnerRoleNameSuffix, preparedDbName)
```
The operator can change owners of DBs listed under `databases`. I think, that already qualifies as revoking privileges. Altering default privileges works additive. Existing privileges won't be revoked.

Note, I moved this part into the function which creates the prepared databases, so that it's not executed on every snyc.

FxKu commented:
1. fixed
2. logs should be clearer now. There was indeed a bug, that roles were not created on UPDATE. Putting an existing database under the prepared section will fail if the owner is different. But, the owner can be changed under `databases` to fix this. Might be worth another paragraph in the docs.
3. I removed the word `bootstrapped` from the docs.

Jan-M commented on docs/user.md:
```diff
@@ -216,6 +219,129 @@ to choose superusers, group roles, [PAM configuration](https://github.com/CyberD
+
+| Role name           | Member of      | Admin         |
+| ------------------- | -------------- | ------------- |
+| foo_owner_user      | foo_owner      | admin         |
```
Yeah that is a bit difficult. Ideally not the case. But then often by mistake objects are ultimately created and owned by the login role. @CyberDem0n any thoughts?

sdudoladov commented:
> But, the owner can be changed under databases to fix this. Might be worth another paragraph in the docs.

As a further note, everything Sync touches after prepared DBs (currently connection pooler), will not be Synced until permission issues are resolved

sdudoladov commented:
:+1:

FxKu commented:
:+1:
